lib/services/role.go

```diff
@@ -596,17 +1127,60 @@ func (set RoleSet) String() string {
 	return fmt.Sprintf("roles %v", strings.Join(roleNames, ","))
 }
 
-// CheckResourceAction checks if role set has access to this resource action
-func (set RoleSet) CheckResourceAction(resourceNamespace, resourceName, accessType string) error {
-	resourceNamespace = ProcessNamespace(resourceNamespace)
+func (set RoleSet) checkAccessToRule(namespace string, resource string, verb string) error {
+	var errs []error
+
+	// check deny: a single match on a deny rule prohibits access
+	for _, role := range set {
+		matchNamespace := MatchNamespace(role.GetNamespaces(Deny), ProcessNamespace(namespace))
+		matchRule := MatchRule(role.GetRules(Deny), resource, verb)
+		if matchNamespace && matchRule {
+			return trace.AccessDenied("role %v denied %v access to %v in namespace %v: deny rule matched", role, verb, resource, namespace)
+		}
+	}
+
+	// check allow: if rule matches, grant access to resource
+	for _, role := range set {
+		matchNamespace := MatchNamespace(role.GetNamespaces(Allow), ProcessNamespace(namespace))
+		matchRule := MatchRule(role.GetRules(Allow), resource, verb)
+		if matchNamespace && matchRule {
+			return nil
+		}
+
+		errorMessage := fmt.Sprintf("role %v denied %v access to %v in namespace %v: no allow rule match", role, verb, resource, namespace)
+		errs = append(errs, trace.AccessDenied(errorMessage))
+	}
+
+	return trace.AccessDenied("%v access to %v denied to role(s): %v: %v", verb, resource, set, errs)
+}
+
+func (set RoleSet) checkAccessToResource(namespace string, resource string, verb string) error {
+	namespace = ProcessNamespace(namespace)
 	for _, role := range set {
-		matchNamespace := MatchNamespace(role.GetNamespaces(), resourceNamespace)
-		matchResourceAction := MatchResourceAction(role.GetResources(), resourceName, accessType)
-		if matchNamespace && matchResourceAction {
+		matchNamespace := MatchNamespace(role.GetNamespaces(Allow), namespace)
+		matchSystemResource := MatchRule(role.GetSystemResources(Allow), resource, verb)
+		if matchNamespace && matchSystemResource {
 			return nil
 		}
 	}
-	return trace.AccessDenied("%v access to %v in namespace %v is denied for %v", accessType, resourceName, resourceNamespace, set)
+	return trace.AccessDenied("%v access to %v in namespace %v is denied for %v", verb, resource, namespace, set)
+}
+
+// CheckAccessToRuleOrResource checks if role set has access to this resource action
+func (set RoleSet) CheckAccessToRuleOrResource(namespace string, resource string, verb string) error {
+	// always check high level rules first
+	ruleErr := set.checkAccessToRule(namespace, resource, verb)
+	if ruleErr != nil {
+		// check low level resource if rules didn't match
+		resourcesErr := set.checkAccessToResource(namespace, resource, verb)
+		if resourcesErr != nil {
+			errorMessage := fmt.Sprintf("unable to access rule: %v and resource: %v", ruleErr, resourcesErr)
```

Comment:
users don't know any thing about system resources, so error should only refer to rule error